community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/IndexSpecifier.java

```diff
@@ -58,21 +58,31 @@ public String toString()
         return specification;
     }
 
-    private Pair<String, String> parse()
+    private Pair<String,String[]> parse()
     {
-        Pattern pattern = Pattern.compile( "" +
-                ":" + or( simpleIdentifier( "simpleLabel" ), complexIdentifier( "complexLabel" ) ) +
-                "\\(" + or( simpleIdentifier( "simpleProperty" ), complexIdentifier( "complexProperty" ) ) + "\\)"
+        Pattern pattern = Pattern.compile(
+                ":" + or( identifier(), qoutedIdentifier() ) + // Match the label
+                "\\(" + or( identifier(), qoutedIdentifier() ) + // Match the first property
+                "(?:\\,\\s" + or( identifier(), qoutedIdentifier() ) + ")*\\)" // Match following properties
         );
         Matcher matcher = pattern.matcher( specification );
         if ( !matcher.find() )
         {
             throw new IllegalArgumentException( "Cannot parse index specification " + specification );
         }
-        return Pair.of(
-                either( matcher.group( "simpleLabel" ), matcher.group( "complexLabel" ) ),
-                either( matcher.group( "simpleProperty" ), matcher.group( "complexProperty" ) )
-        );
+        int group = 0;
+        String label = either( matcher.group( ++group ), matcher.group( ++group ) );
+        ArrayList<String> properties = new ArrayList<>();
+
+        while ( group + 2 <= matcher.groupCount() )
+        {
+            String property = either( matcher.group( ++group ), matcher.group( ++group ) );
+            if ( property != null )
+            {
+                properties.add( property );
+            }
+        }
+        return Pair.of( label, properties.toArray( new String[0] ) );
```

Comment:
This code is not easy to follow. It is tightly coupled with how the capturing and non-capturing groups are organised in the regular expression. Some comments would be nice.